// vite-plugin-blinko/dist/index.js
import * as fs from 'fs';
import * as path from 'path';
// Default required fields for plugin.json validation
const DEFAULT_REQUIRED_FIELDS = ['name', 'author', 'url', 'version', 'minAppVersion', 'displayName', 'description'];
/**
 * Generates a random string for development build filenames
 * @param {number} length - Length of the random string
 * @returns {string} Random string
 */
const generateRandomString = (length = 8) => {
    return Math.random().toString(36).substring(2, length + 2);
};
/**
 * Vite plugin for Blinko application development
 * @param options - Plugin options
 * @returns Vite plugin
 */
export default function blinkoPlugin(options = {}) {
    const { entry = 'src/index.tsx', prodOutDir = 'release', devOutDir = 'dist', requiredFields = DEFAULT_REQUIRED_FIELDS, pluginJsonPath = path.resolve(process.cwd(), 'plugin.json'), externals = ['blinko'], useRandomFilename = true } = options;
    return [
        {
            name: 'vite-plugin-blinko-check-json',
            buildStart() {
                if (!fs.existsSync(pluginJsonPath)) {
                    throw new Error(`plugin.json not found at path: ${pluginJsonPath}`);
                }
                try {
                    const pluginData = JSON.parse(fs.readFileSync(pluginJsonPath, 'utf-8'));
                    const missingFields = requiredFields.filter(field => !(field in pluginData));
                    if (missingFields.length > 0) {
                        throw new Error(`Missing required fields in plugin.json: ${missingFields.join(', ')}`);
                    }
                }
                catch (error) {
                    if (error instanceof SyntaxError) {
                        throw new Error('Invalid JSON in plugin.json');
                    }
                    throw error;
                }
            }
        },
        {
            name: 'vite-plugin-blinko-config',
            config(config, { mode }) {
                const isProd = mode === 'production';
                return {
                    build: {
                        lib: {
                            entry,
                            formats: ['es'],
                            fileName: (!isProd && useRandomFilename)
                                ? () => `index_${generateRandomString()}.js`
                                : 'index'
                        },
                        outDir: isProd ? prodOutDir : devOutDir,
                        rollupOptions: {
                            external: externals,
                            output: {
                                inlineDynamicImports: true,
                                banner: '(function() {',
                                footer: '})();',
                            }
                        }
                    },
                    json: {
                        stringify: false,
                    }
                };
            }
        }
    ];
}
